sql: correctly handle COLLATEd columns in COPY

otan opened the pull request:
See individual commits for details.

Resolves #95887

Release justification: fixes a critical bug with COPY

otan committed:
tree: rename ParseTimeContext to ParseContext

A mechanical rename PR. Anything named after ParseTimeContext is now
just ParseContext.

Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/95894)
<!-- Reviewable:end -->

cucaroach reviewed (commented):
What was the error?   Might be good to mention it in the commit notes in case it happens again.

otan committed:
sql: fix collated strings during COPY

Release note (bug fix): Fix a bug where COPYing into a column with
collated strings would result in an error similar to
`internal error: unknown type collatedstring`.

otan commented:
done! tftr!

bors r=cucaroach
